### PR TITLE
[flang][OpenMP] Parsing support for iterator in DEPEND clause

### PR DESCRIPTION
**Pull request by kparzysz**
Warn about use of iterators OpenMP versions that didn't have them (support added in 5.0). Emit a TODO error in lowering.

### CODE REVIEW

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 1]
```diff
@@ -3307,6 +3307,15 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Depend &x) {
         }
       }
     }
+    if (std::get<std::optional<parser::OmpIteratorModifier>>(inOut->t)) {
+      unsigned version{context_.langOptions().OpenMPVersion};
+      unsigned allowedInVersion = 50;
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3312]:
```suggestion
      unsigned allowedInVersion{50};
```

**Review by kiranchandramohan — APPROVED** [review]
LG.

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 2]
```diff
@@ -795,35 +795,41 @@ bool ClauseProcessor::processCopyprivate(
...
-                 "substring not supported for task depend");
-          } else if (evaluate::IsArrayElement(*object.ref())) {
-            TODO(converter.getCurrentLocation(),
-                 "array sections not supported for task depend");
-          }
+  auto process = [&](const omp::clause::Depend &clause,
+                     const parser::CharBlock &) {
+    using Depend = omp::clause::Depend;
+    assert(std::holds_alternative<Depend::DepType>(clause.u) &&
+           "Only the form with depenence type is handled at the moment");
```
Comment (tblah) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:802]: Should this be a TODO?

Comment (kparzysz) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp]: Done.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp]: Done

**Pull request merged**